File: Treachery-iOS/Treachery-iOS/Models/IdentityCard.swift
//
//  IdentityCard.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/12/26.
//

import Foundation

struct IdentityCard: Codable, Identifiable, Hashable {
    let id: String
    let cardNumber: Int
    let name: String
    let role: Role
    let abilityText: String
    let unveilCost: String
    let rarity: Rarity
    let hasUndercover: Bool
    let undercoverCondition: String?
    let timingRestriction: String?
    let lifeModifier: Int?
    let handSizeModifier: Int?
    let flavorText: String?
    let imageAssetName: String?

    enum CodingKeys: String, CodingKey {
        case id
        case cardNumber = "card_number"
        case name
        case role
        case abilityText = "ability_text"
        case unveilCost = "unveil_cost"
        case rarity
        case hasUndercover = "has_undercover"
        case undercoverCondition = "undercover_condition"
        case timingRestriction = "timing_restriction"
        case lifeModifier = "life_modifier"
        case handSizeModifier = "hand_size_modifier"
        case flavorText = "flavor_text"
        case imageAssetName = "image_asset_name"
    }
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/model/IdentityCard.kt
package com.solomon.treachery.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class IdentityCard(
    val id: String,
    @SerialName("card_number") val cardNumber: Int,
    val name: String,
    val role: String,
    @SerialName("ability_text") val abilityText: String,
    @SerialName("unveil_cost") val unveilCost: String,
    val rarity: String,
    @SerialName("has_undercover") val hasUndercover: Boolean,
    @SerialName("undercover_condition") val undercoverCondition: String? = null,
    @SerialName("timing_restriction") val timingRestriction: String? = null,
    @SerialName("life_modifier") val lifeModifier: Int? = null,
    @SerialName("hand_size_modifier") val handSizeModifier: Int? = null,
    @SerialName("flavor_text") val flavorText: String? = null,
    @SerialName("image_asset_name") val imageAssetName: String? = null
) {
    val roleEnum: Role? get() = Role.fromValue(role)
    val rarityEnum: Rarity? get() = Rarity.fromValue(rarity)
}
